rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,387 @@
+The advantage with this approach is that it is very flexible, allowing us to stub functions, methods, and types, either by directly replacing them, or appending their replacements and injecting appropriate conditional compilation guards.
+
+This approach entails less user effort than Alternative #1, but it has the same downside that it requires all source code to be available.
+It also might be difficult to inject code in a way that names are correctly resolved (e.g., if the replacement code comes from a different crate).
+Also, source code is difficult to work with (e.g., unexpanded macros).
+
+On the last two points, we might be able to take advantage of an existing source analysis platform like `rust-analyzer` (which has facilities like structural search replace), but this would add more (potentially fragile) dependencies to Kani.
+
+### Alternative #3: AST-to-AST or HIR-to-HIR transformation
+
+In this alternative, we implement stubbing by rewriting the [AST](https://rustc-dev-guide.rust-lang.org/syntax-intro.html) or [High-Level IR (HIR)](https://rustc-dev-guide.rust-lang.org/hir.html) of the program.
+The HIR is a more compiler-friendly version of the AST; it is what is used for type checking.
+To swap out a function, method, or type at this level, it looks like it would be necessary to add another pass to `rustc` that takes the initial AST/HIR and produces a new AST/HIR with the appropriate replacements.
+
+The advantage with this approach is, like source transformations, it would be very flexible.
+The downside is that it would require modifying `rustc` (as far as we know, there is not an API for plugging in a new AST/HIR pass), and would also require performing the transformations at a very syntactic level: although the AST/HIR would likely be easier to work with than source code directly, it is still very close to the source code and not very abstract.
+Furthermore, it would require that we have access to the AST/HIR for all external code, and---provided we supported stubbing across crate boundaries---we would need to figure out how to inject the AST/HIR from one crate into another (the AST/HIR is usually just constructed for the crate currently being compiled).
+
+## Open questions
+
+- Would there ever be the need to stub a particular monomorphization of a function, as opposed to the polymorphic function?
+- What does it mean for the replacement function/method to be "compatible" with the original one?
+Requiring the replacement's type to be a subtype of the original type is likely stronger than what we want.
+For example, if the original function is polymorphic but monomorphized to only a single type, then it seems okay to replace it with a function that matches the monomorphized type.
+- How can we allow a user to stub a method and access private fields of `self`?
+This should be possible using `std::mem::transmute`, but can we hide this from the user and make it less brittle?
+
+## Future possibilities
```

Comment:
Should we add a limitations section? E.g.:
- This won't work with `--concrete-playback` for now.
- We are only able to apply abstractions to dependencies if user enables MIR Linker.